### PR TITLE
Update gatk4spark/markduplicates with nftest

### LINKED ISSUE #4615: [FEATURE] gatk4spark/markduplicates nf-test
### Is your feature request related to a problem? Please describe

gatk4spark/markduplicates lacks nf-tests

### Describe the solution you'd like

add nf-tests, remove pytests

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by anoronh4**
## PR checklist

Closes #4615 
also added a stub block.

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Review by SPPearce — COMMENTED** [review]
Tests look good, just have concerns about how the file name extension is currently being set. I realise that isn't your change here, but If no `prefix` is specified, this will make a file with no extension just called `meta.id`

File: modules/nf-core/gatk4spark/markduplicates/main.nf [hunk 1]
```diff
@@ -46,7 +46,18 @@ process GATK4SPARK_MARKDUPLICATES {
...
     "${task.process}":
         gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
-        openjdk: \$(echo \$(java -version 2>&1) | grep version | sed 's/\"//g' | cut -f3 -d ' ')
+    END_VERSIONS
+    """
+
+    stub:
+    prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}
```
[the remark this answers is not in the document]
Comment (anoronh4) [reply modules/nf-core/gatk4spark/markduplicates/main.nf:55]: i fixed the logic for both the script block and the stub block.

**Comment by maxulysse** [comment]
You're killing it @anoronh4 thank you so much!!!

File: modules/nf-core/gatk4spark/markduplicates/tests/main.nf.test.snap [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    "homo_sapiens - cram": {
+        "content": [
+            {
+                "0": [
+                    [
+                        {
+                            "id": "test",
+                            "single_end": false
+                        },
+                        "test.bam:md5,898cb0a6616897d8ada90bab53bf0837"
```
Comment (SPPearce) [inline modules/nf-core/gatk4spark/markduplicates/tests/main.nf.test.snap:121]: Why is this a bam, shouldn't it be a cram?

File: modules/nf-core/gatk4spark/markduplicates/main.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (SPPearce) [reply modules/nf-core/gatk4spark/markduplicates/main.nf:55]: So this now makes bam files unless explicitly asked for cram files? 
Is the cram test just missing the config (I haven't used cram files so not sure how they really work)

File: modules/nf-core/gatk4spark/markduplicates/tests/main.nf.test.snap (hunk 2 above)
Comment (anoronh4) [reply modules/nf-core/gatk4spark/markduplicates/tests/main.nf.test.snap]: yes! thank you for catching.

File: modules/nf-core/gatk4spark/markduplicates/main.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (anoronh4) [reply modules/nf-core/gatk4spark/markduplicates/main.nf:55]: yes, that's what i've set the default behavior as. there is no argument in the command, the only way to indicate the output format is by adding the extension you want in the output. we could change this behavior by adding another input channel, but this does force users to change how the module is implemented in their pipeline.

File: modules/nf-core/gatk4spark/markduplicates/tests/main.nf.test.snap (hunk 2 above)
Comment (anoronh4) [reply modules/nf-core/gatk4spark/markduplicates/tests/main.nf.test.snap]: the cram test was pointing to a config but the config had the wrong process selector name in it...it was carried over from the pytests, which allowed aliases. i have removed the alias and now the config is being correctly applied. cram is showing up. let me know if there's anything else!

**Review by SPPearce — APPROVED** [review]

**Pull request merged**